server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -661,15 +693,46 @@ public ParsedDocument index(IndexItem item) throws IOException {
         }
     }
 
+    /**
+     *
+     * RETURNING can refer to multiple non-deterministic columns,
+     * generated and defaults and their combinations, e.g., gen_col + def_col.
+     *
+     * We need to ensure that we use already computed values for such references,
+     * this this must be called after {@link #index(IndexItem)} have been called.
+     *
+     */
     @Nullable
     public Object[] returnValues(IndexItem item) {
-        if (this.returnValueInputs == null) {
+        if (this.returnValues == null) {
             return null;
         }
+
+        // We must not change original RETURNING symbols.
+        // Otherwise, RefReplacer.replaceRefs will changes Refs to Literals on the first index()
+        // and all subsequent calls will not find a Reference and will operate with old Literals.
+        Symbol[] updatedReturnValues = new Symbol[returnValues.length];
+
+        for (int i = 0; i < returnValues.length; i++) {
+            Symbol updatedReturnValue = RefReplacer.replaceRefs(returnValues[i], ref -> {
+                Synthetic synthetic = synthetics.get(ref.column());
+                if (synthetic != null) {
+                    // Deterministic values are also reused just in order not to compute twice.
+                    return Literal.ofUnchecked(ref.valueType(), synthetic.computedValue());
+                }
+                return ref;
+            });
+            updatedReturnValues[i] = updatedReturnValue;
+        }
+
+        List<Input<?>> returnValueInputs = new ArrayList<>(returnValues.length);
+        for (Symbol returnValue : updatedReturnValues) {
+            returnValueInputs.add(ctxForRefs.add(returnValue));
+        }
```

Comment:
I suspect re-creating inputs for the return values on each row may have quite a bit of impact on performance.

Couldn't we let `Synthetic` implement `Input` and use them within the `returnValueInputs`, so we don't need to re-create them per row?

That should also allow to fully encapsulate the cache on a `value()` call. Right now consumers have to know when to call `value()` and when to call `computedValue()` or activate the cache.

Downside is that we'd have to bring back invalidation.

Reply:
Thanks for the suggestion!

Implemented in https://github.com/crate/crate/pull/14762/commits/529b01e17ca57bdbc1a80bc601883ed6dec0b3ea